### ISSUE TITLE
[Android] tapping push notification does nothing when app not running or in background

### ISSUE THREAD

**Issue by CFT-Chris** [ionic-team/capacitor#1475]
Description of the problem:

Affected platform
- [x] Android
- [ ] iOS
- [ ] electron
- [ ] web

OS of the development machine
- [x] Windows
- [ ] macOS
- [ ] linux

Other information: This is the same as issue #1192, which was closed under the impression it was fixed with the latest beta (22). It still reproduces for me, but I have investigated it and am putting in a PR very soon.

Capacitor version: beta.22

node version: 10.13

npm version: 6.4

CocoaPods version: n/a

Steps to reproduce:

Send Firebase (FCM) message to Capacitor app not running in foreground. Tap on notification in Android's notification panel.

Either nothing happens (app is not started or background app not brought to foreground), or the app is started up without sending the notification data to any of its listeners `pushNotificationReceived` or `pushNotificationActionPerformed`.

Link to sample project: n/a

**Comment by jcesarmobile** [comment]
Can you show your notification payload? It works fine for me without all the changes in your PR

**Comment by CFT-Chris** [comment]
Thanks for the quick reply. 

I am testing on Android 5.0 (minimum required by Capacitor).

I tried all payload combinations.  Without "notification", then without "data", with both "notification" and "data", with "click_event" and without, etc.

Just an example of what I am sending (via Node.js backend POST request to Firebase cloud) which works with the changes I proposed in the PR:
```
request({
  url: 'https://fcm.googleapis.com/fcm/send',
  method: 'POST',
  headers: {
    'Content-Type' :' application/json',
    'Authorization': 'key=' + /* FCM key */
  },
  body: JSON.stringify({
    to: / *** PushReg Id ***/,
    data: { tm: Date.now() },
    notification: {
      sound: 'default',
      click_action: "CAPACITOR_FIREBASE_MESSAGING_ACTIVITY"
    }
  }
})
```

Without "notification" the notification doesn't even arrive at all if the app is not running (background or killed).

Just looking at the code itself, I can't figure out how Android ends up calling [handleOnNewIntent in PushNotifications.java](https://github.com/ionic-team/capacitor/blob/master/android/capacitor/src/main/java/com/getcapacitor/plugin/PushNotifications.java#L59) as there is nothing in the AndroidManifests, apart from `CapacitorFirebaseMessagingService`, that might deal with incoming Firebase messages when the app is not running in the foreground.

In my case, when the Capacitor app is in the background or not running, the Firebase message results in a notification in Android's notification panel, and tapping that notification does not make CapacitorFirebaseMessagingService's `onMessageReceived` handle it. 

Could be old Android behaviour that doesn't repro in later OS versions, but regardless, how would the code inside handleOnNewIntent (pushnotifications.java) get called, and how does any listener to `pushNotificationActionPerformed` get anything?  Do later versions of Android somehow know to hook into the PushNotifications plugin and send an Intent there?

Sorry if I overlooked something obvious.

**Comment by CFT-Chris** [comment]
I guess I should also elaborate that I have just read #1349 and noticed that @jurepurgar mentioned that for Android, his fix is partial and provides a step for the changes I'm proposing. 

Inadvertently, I think I may have come up with the other half of the fix (setting MainActivity to `singleTask`, sending the data if the app is closed once a listener is added, etc.), and the way I get it to call his onNewIntent code for Android is to use the "click_action", which is a method employed by the Cordova FCM plugin for Ionic Native.

And one more thing, my payload example above omitted "title" and "body", but they are there. I mistakenly edited them out when copying from my code and pasting to here.

**Comment by jcesarmobile** [comment]
Supposedly, if there is no click action it should open the BridgeActivity, and BridgeActivity have the code to handle the intent, it pass the data to the Bridge and the bridge send it to the plugins. But will double check tomorrow

**Comment by CFT-Chris** [comment]
Ok, that info is very helpful.  I didn't realize that MainActivity eventually had its Intents routed to `handleOnNewIntent` in the PushNotifications plugin. I think I can accurately narrow down the cases that fail for me.

**tl;dr**: The notification in the payload must be present, but click_action must not be in the notification!

***Sample FCM payloads:***
**Bad**: { data: { title: 'foo', body: 'bar' } }
**Bad**: { data: { title: 'foo', body: 'bar' }, notification: { click_action: 'FCM_PLUGIN_ACTIVITY' } }
**Good**: { data: { title: 'foo', body: 'bar' }, notification: {} }
**Good**: { data: { title: 'foo', body: 'bar' }, notification: { sound: true } }

I still think `android:launchMode="singleTop"` (anything but `standard`) is necessary for MainActivity.

Here were my test cases and the results (Android 5.0):

- **Without** `android:launchMode="singleTop"`, and:
  - App in foreground: `pushNotificationReceived` fired (works)
  - App **not** in foreground, and:
    - No notification in payload: no notification to tap (expected, my mistake for assuming it would leave a notification — I read it somewhere on stackoverflow and is either bad info or I misinterpreted it)
    - Notification in payload with click_action: nothing happens (expected, my mistake again for not noticing that I was sending an unused click_action with my initial payloads when I migrated to Capacitor from Ionic Native plugins)
    - Notification in payload without click_action: notification when tapped either (1) destroys app when in background, restarts app with notification data *or* (2) another webview appears on top of the first one, and the first instance gets the notification data but the second and now active one does not (the desired behaviour would simply to have the existing instance get the notification data)
- **With** `android:launchMode="singleTop"`, and:
  - App in foreground: `pushNotificationReceived` fired (works)
  - App **not** in foreground, and:
    - No notification in payload: no notification to tap (expected)
    - Notification in payload without click_action: `pushNotificationActionPerformed` fired and handled by the existing app instance (works!)

I'm glad that's all cleared up now!  Still would like to push a PR though to change the launchMode to `singleTop` to prevent the Ionic app restarting as a result of a new activity being created every time a push notification is tapped.  As a result of recreating MainActivity we get the problems of re-init'ing/re-loading BridgeActivity leading to the bad behaviour described above.

**Comment by solojuve1897** [comment]
Does this mean the solution is using singleTop and that other options like singleTask or singleInstance doesn't work fully? Because I'm pretty sure I have some of the issues you are referring to even though I'm using singleTask. Need to test this more today. Will write a comment about me results.

**Comment by solojuve1897** [comment]
My bad. Works fine and I'm using singleTask. Would be good to know though which one is best and why: singleTop, singleInstance or singleTask.

**Comment by CFT-Chris** [comment]
Sounds good to me.  I couldn't think of common scenarios where `singleTop` would be preferred, but I think any apps that desire such behavior are free to change the `launchMode` in AndroidManifest.xml manually.

PR is updated.

**Comment by jcesarmobile** [comment]
Fixed by https://github.com/ionic-team/capacitor/pull/1477

**Comment by vadimrostok** [comment]
We are moving to Capacitor and needed to support old and new systems, so notifications contained `click_action: ST_ACTIVATE` field.
The solution was to add following intent-filter to main activity in AndroidManifest.xml:
```xml
<intent-filter>
  <action android:name="ST_ACTIVATE" />
  <category android:name="android.intent.category.DEFAULT" />
</intent-filter>
```
